list_updatable_packages

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python3
+
+import json
+import os
+import subprocess
+import sys
+from pathlib import Path
+from typing import Generator, Iterable, Mapping, Optional
+
+import requests
+
+SUPPORTED_TEMPLATES = ('foreman_plugin', 'hammer_plugin', 'smart_proxy_plugin')
+SESSION = requests.Session()
+
+
+class Spec:
+    def __init__(self, path: Path):
+        self.path = path
+        spec = path.read_text()
+        self.lines = spec.splitlines()
+        self._globals = None
+
+    @property
+    def directory(self) -> str:
+        return self.path.parent.as_posix()
+
+    @property
+    def package_name(self) -> str:
+        return self.path.name[:-len('.spec')]
+
+    @property
+    def template(self) -> Optional[str]:
+        for line in self.lines:
+            if line.startswith('# template:'):
+                _, _, template = line.split(None, 3)
+                return template
+
+        return None
+
+    @property
+    def globals(self) -> Mapping[str, str]:
+        """
+        Perform very naive spec parsing to extract the globals
+        """
+        if self._globals is None:
+            self._globals = {}
+            for line in self.lines:
+                if line.startswith('%global'):
+                    definition, name, value = line.split(None, 3)
+                    if definition == '%global':
+                        self._globals[name] = value
+
+        return self._globals
+
+    @property
+    def is_nightly(self) -> bool:
+        return 'prereleasesource' in self.globals
+
+    @property
+    def gem_name(self) -> str:
+        return self.globals.get('gem_name')
+
+    @property
+    def version(self) -> str:
+        command = ['rpmspec', '--srpm', '-q', '--queryformat=%{version}', self.path.as_posix()]
+        return subprocess.check_output(command, stderr=subprocess.DEVNULL, universal_newlines=True)
```

Comment:
This part is slow but I'm not sure if there's a faster way (that's still easily maintainable). AFAIK rpm isn't really a nice library to use though it would probably be faster.

Reply:
`rpmspec`: `./list_updatable_packages  2.24s user 0.48s system 8% cpu 32.325 total`
`import rpm`: `./list_updatable_packages  1.93s user 0.08s system 9% cpu 20.501 total`

I wouldn't bother too much ;)

Reply:
```diff
diff --git list_updatable_packages list_updatable_packages
index fd637781a..33ef64fb0 100755
--- list_updatable_packages
+++ list_updatable_packages
@@ -8,6 +8,7 @@ from pathlib import Path
 from typing import Generator, Iterable, Mapping, Optional
 
 import requests
+import rpm
 
 SUPPORTED_TEMPLATES = ('foreman_plugin', 'hammer_plugin', 'smart_proxy_plugin')
 SESSION = requests.Session()
@@ -62,8 +63,9 @@ class Spec:
 
     @property
     def version(self) -> str:
-        command = ['rpmspec', '--srpm', '-q', '--queryformat=%{version}', self.path.as_posix()]
-        return subprocess.check_output(command, stderr=subprocess.DEVNULL, universal_newlines=True)
+        ts=rpm.TransactionSet()
+        s = ts.parseSpec(self.path.as_posix())
+        return s.sourceHeader.format('%{version}')
 
 
 def find_specs() -> Generator[Spec, None, None]:
```

if you're trully interested.

Reply:
You'd only take the `import rpm` overhead once instead of once per package. I'll see about the entire runtime.

Reply:
`rpmspec` is not Python, it's C.

Toying with the Debian side of things, the slow part of this all is talking to `rubygems.org`, not parsing specs.

Reply:
with calling rpmspec:
```console
$ time ./list_updatable_packages 
packages/plugins/rubygem-smart_proxy_reports 0.0.7
packages/plugins/rubygem-smart_proxy_salt 4.0.0
packages/plugins/rubygem-smart_proxy_dhcp_dnsmasq 1.0
packages/plugins/rubygem-foreman_salt 14.1.0
packages/plugins/rubygem-smart_proxy_dhcp_bluecat 0.1.6

real	0m29.179s
user	0m3.772s
sys	0m0.964s
```

With using RPM as a library:
```console
$ time ./list_updatable_packages 
warning: line 21: It's not recommended to have unversioned Obsoletes: Obsoletes: rubygem-redhat_access
warning: line 22: It's not recommended to have unversioned Obsoletes: Obsoletes: rubygem-redhat_access_lib
warning: line 24: It's not recommended to have unversioned Obsoletes: Obsoletes: rubygem-foreman_inventory_upload
warning: line 25: It's not recommended to have unversioned Obsoletes: Obsoletes: rubygem-foreman_inventory_upload-doc
warning: line 21: It's not recommended to have unversioned Obsoletes: Obsoletes: rubygem-redhat_access
warning: line 22: It's not recommended to have unversioned Obsoletes: Obsoletes: rubygem-redhat_access_lib
warning: line 24: It's not recommended to have unversioned Obsoletes: Obsoletes: rubygem-foreman_inventory_upload
warning: line 25: It's not recommended to have unversioned Obsoletes: Obsoletes: rubygem-foreman_inventory_upload-doc
packages/plugins/rubygem-smart_proxy_reports 0.0.7
packages/plugins/rubygem-smart_proxy_salt 4.0.0
packages/plugins/rubygem-smart_proxy_dhcp_dnsmasq 1.0
packages/plugins/rubygem-foreman_salt 14.1.0
packages/plugins/rubygem-smart_proxy_dhcp_bluecat 0.1.6

real	0m24.668s
user	0m3.658s
sys	0m0.175s
```

So about 15% faster to use it as a library, but I'd need to look at silencing the warnings. I'd be in favor of keeping it as it is now and only look at it if we really need to.

Reply:
Same. Please merge ;)